### PR TITLE
build MPI

### LINKED ISSUE #132: MPI builds
### Comment:

It seems to me that MPI is not enabled in this recipe. I'd like to know if there is any specific reason for it, or if I can contribute to enable MPI.

### PR DESCRIPTION
**Pull request by njzjz**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

Fix #132.

### CODE REVIEW

**Comment by njzjz** [comment]
@conda-forge-admin, please rerender

**Comment by njzjz** [comment]
@conda-forge-admin, please rerender

File: recipe/meta.yaml
Comment (mikemhenry) [inline recipe/meta.yaml:8]: What is the reason to have a different build number for mpi/nompi?

Comment (njzjz) [reply recipe/meta.yaml:8]: Soery it should be mpi=='nompi'. Per https://github.com/conda-forge/ambertools-feedstock/issues/132#issuecomment-1809073915 users may prefer nompi

Comment (mikemhenry) [reply recipe/meta.yaml:8]: I mean, why do the `nompi` and `mpi` have/need different build numbers? Is that convention for `mpi` enabled/aware packages on conda-forge? 

You also might want to try and get this working on the linux builds first and skip the other builds to save you some time getting things to pass on the CI. Working locally might be a good idea as well `python build-locally.py` since it will be faster than the azure CI. Thanks for taking this on!

Comment (njzjz) [reply recipe/meta.yaml:8]: > Is that convention for mpi enabled/aware packages on conda-forge?

It is explained in the documentation: https://conda-forge.org/docs/maintainer/knowledge_base.html

> Up to here, mpi providers have no explicit preference. When choosing an MPI provider, the mutual exclusivity of the mpi metapackage allows picking between mpi providers by installing an mpi provider, e.g.
>
> conda install mpich ptscotch
or
>
> conda install openmpi ptscotch
This doesn’t extend to nompi, because there is no nompi variant of the mpi metapackage. And there probably shouldn’t be, because some packages built with mpi don’t preclude other packages in the env that may have an mpi variant from using the no-mpi variant of the library (e.g. for a long time, fenics used mpi with no-mpi hdf5 since there was no parallel hdf5 yet. This works fine, though some features may not be available).
>
> Typically, if there is a preference it will be for the serial build, such that installers/requirers of the package only get the mpi build if explicitly requested. We use a higher build number for the nompi variant in this case.

Comment (mattwthompson) [reply recipe/meta.yaml:8]: Using build numbers to trick solvers into preferring particular variants is unfortunately pervasive. I introduced it into GROMACS as a way of keeping single precision the default when I added double precision builds. I'd be nice to use a better solution if one exists - the documentation is correct but could be out of date - I haven't looked into it lately

File: recipe/meta.yaml
Comment (jaimergp) [inline recipe/meta.yaml:26]:
```suggestion
  skip: True  # [win or py2k or ppc64le or aarch64 or py!=39]
```

Do note the explosion of items in the job matrix. Might be preferable to keep it to a single python version while iterating. Add this and rerender if you agree.

File: recipe/meta.yaml
Comment (mikemhenry) [reply recipe/meta.yaml:8]: Sounds good! I am fine keeping `nompi` the default flavor

File: recipe/meta.yaml
Comment (mikemhenry) [reply recipe/meta.yaml:26]: +1 While this can be done locally, at the end of the day it will require some ping-pong with the CI, so turning off builds you don't need and working on one at a time is how I like to do it

File: recipe/meta.yaml
Comment (njzjz) [reply recipe/meta.yaml:8]:
```suggestion
{% set build = build + 100 %}  # [mpi == 'nompi']
```

File: recipe/meta.yaml
Comment (njzjz) [inline recipe/meta.yaml:27]: Revert after everything is fine:
```suggestion
  skip: True  # [win or py2k or ppc64le or aarch64]
```

**Review by njzjz — COMMENTED** [review]
Now it passes on my computer after multiple retries. I am going to see if it passes here

**Comment by njzjz** [comment]
Got this error for osx-arm64:

```
2024-01-03T08:16:13.8739220Z + conda install --yes --no-deps --force-reinstall -p /Users/runner/miniforge3/conda-bld/ambertools_1704269182095/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol xorg-xproto xorg-libx11
2024-01-03T08:16:13.8769280Z + local cmd=install
2024-01-03T08:16:13.8842690Z + case "$cmd" in
2024-01-03T08:16:13.8871280Z + __conda_exe install --yes --no-deps --force-reinstall -p /Users/runner/miniforge3/conda-bld/ambertools_1704269182095/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol xorg-xproto xorg-libx11
2024-01-03T08:16:13.8873080Z + /Users/runner/miniforge3/bin/conda install --yes --no-deps --force-reinstall -p /Users/runner/miniforge3/conda-bld/ambertools_1704269182095/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol xorg-xproto xorg-libx11
2024-01-03T08:16:16.5440100Z Channels:
2024-01-03T08:16:16.5440880Z  - conda-forge
2024-01-03T08:16:16.5441260Z Platform: osx-64
2024-01-03T08:16:50.9159540Z Collecting package metadata (repodata.json): ...working... done
2024-01-03T08:16:51.2541850Z Solving environment: ...working... failed
2024-01-03T08:16:51.2545380Z 
2024-01-03T08:16:51.2546940Z InvalidSpec: The package "conda-forge/osx-arm64::python_abi==3.9=4_cp39" is not available for the specified platform
```

It's not related to this PR but should be resolved.

@jaimergp, could you please clarify why they need to be reinstalled? Should they be installed for the host or target platform? Thanks!

**Comment by njzjz** [comment]
Oh, I see these comments:

https://github.com/conda-forge/ambertools-feedstock/blob/f14f8759cb85823b5bd1a62403e7167be1f4469f/recipe/build.sh#L39-L45

I think it should be reinstalled for the host platform.

**Comment by njzjz** [comment]
All builds pass. Now I'll build for other Python versions.

**Comment by njzjz** [comment]
@conda-forge-admin, please rerender

File: recipe/build.sh
Comment (mattwthompson) [inline recipe/build.sh:49]: I don't understand why this change is needed - could you help me out?

**Comment by njzjz** [comment]
Hi @conda-forge/ambertools, could someone take a review and merge the PR? Everything should be fine.

**Comment by mattwthompson** [comment]
I think this is good to go, I'm not sure if I'm missing something

**Comment by mattwthompson** [comment]
@dacase could you raise a new issue about version 24? It'll either get lost here or hijack this thread

**Comment by xiki-tempula** [comment]
I wonder if there is any update for this PR? This is a very valuable addition to the current conda package as it allows one easily deploy test the MPI version of sander in the unit tests.

**Comment by xiki-tempula** [comment]
@conda-forge-admin, please rerender

**Comment by xiki-tempula** [comment]
Might also be good to add a `sander.MPI -h` to https://github.com/conda-forge/ambertools-feedstock/blob/main/recipe/run_test.sh to check the binary runs.

**Comment by mattwthompson** [comment]
On `osx-arm64` only:

```
  File "/Users/runner/miniforge3/lib/python3.10/site-packages/urllib3/response.py", line 761, in _error_catcher
    raise ProtocolError(arg, e) from e
urllib3.exceptions.ProtocolError: ('Connection broken: IncompleteRead(7551886 bytes read, 562277331 more expected)', IncompleteRead(7551886 bytes read, 562277331 more expected))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/runner/miniforge3/bin/conda-build", line 11, in <module>
    sys.exit(execute())
  File "/Users/runner/miniforge3/lib/python3.10/site-packages/conda_build/cli/main_build.py", line 581, in execute
    api.build(
  File "/Users/runner/miniforge3/lib/python3.10/site-packages/conda_build/api.py", line 250, in build
    return build_tree(
  File "/Users/runner/miniforge3/lib/python3.10/site-packages/conda_build/build.py", line 3762, in build_tree
    packages_from_this = build(
  File "/Users/runner/miniforge3/lib/python3.10/site-packages/conda_build/build.py", line 2547, in build
    try_download(m, no_download_source=False, raise_error=True)
  File "/Users/runner/miniforge3/lib/python3.10/site-packages/conda_build/render.py", line 757, in try_download
    source.provide(metadata)
  File "/Users/runner/miniforge3/lib/python3.10/site-packages/conda_build/source.py", line 1038, in provide
    unpack(
  File "/Users/runner/miniforge3/lib/python3.10/site-packages/conda_build/source.py", line 171, in unpack
    src_path, unhashed_fn = download_to_cache(
  File "/Users/runner/miniforge3/lib/python3.10/site-packages/conda_build/source.py", line 105, in download_to_cache
    download(url, path)
  File "/Users/runner/miniforge3/lib/python3.10/site-packages/conda/exports.py", line 408, in download
    return _download(url, dst_path, md5=md5sum, sha256=sha256, size=size)
  File "/Users/runner/miniforge3/lib/python3.10/site-packages/conda/common/io.py", line 85, in decorated
    return f(*args, **kwds)
  File "/Users/runner/miniforge3/lib/python3.10/site-packages/conda/gateways/connection/download.py", line 66, in download
    download_inner(
  File "/Users/runner/miniforge3/lib/python3.10/site-packages/conda/gateways/connection/download.py", line 122, in download_inner
    for chunk in resp.iter_content(chunk_size=CHUNK_SIZE):
  File "/Users/runner/miniforge3/lib/python3.10/site-packages/requests/models.py", line 818, in generate
    raise ChunkedEncodingError(e)
requests.exceptions.ChunkedEncodingError: ('Connection broken: IncompleteRead(7551886 bytes read, 562277331 more expected)', IncompleteRead(7551886 bytes read, 562277331 more expected))

**Comment by njzjz** [comment]
> On `osx-arm64` only:

This looks like a network issue.

> Might also be good to add a sander.MPI -h to https://github.com/conda-forge/ambertools-feedstock/blob/main/recipe/run_test.sh to check the binary runs.

`sander -h` doesn't return a zero code. Perhaps we could test `which sander` and `which sander.MPI`.

```
sander -h; echo $?

     mdfil: Error unknown flag: -h                                                                                                                                                                                                                                                              

     usage: sander  [-O|A] -i mdin -o mdout -p prmtop -c inpcrd -r restrt
                   [-ref refc -x mdcrd -v mdvel -e mden -frc mdfrc -idip inpdip -rdip rstdip -mdip mddip 
                   -inf mdinfo -radii radii -y inptraj -amd amd.log -scaledMD scaledMD.log] -cph-data -ce-data <file>-host ipi_hostname -port ipi_port
Consult the manual for additional options.
1
```

**Comment by mattwthompson** [comment]
> This looks like a network issue.



Yeah, somewhat suspicious that it's only on one architecture ... maybe the difference is when those runs started?

> Perhaps we could test ...

Sorry to be unclear, that file could be improved by going through those commands and replacing them with something that runs like `sander --version` etc. I think `which sander` etc. would also be fine

I also don't think those changes are necessary to get this merged! They are "nice to have"s

**Comment by mattwthompson** [comment]
@conda-forge-admin, please restart ci

**Comment by mattwthompson** [comment]
The build seem (?) to be getting past the download step, so maybe they're going to succeed

**Comment by xiki-tempula** [comment]
Yes, `sander --version` seems to be a good test.

**Comment by mattwthompson** [comment]
This is probably fine but I want #137 and #139 merged upstream and back into this branch before merging this one

**Comment by mattwthompson** [comment]
```
+ sander --version
sander: Version 22.0
+ sander.MPI --version
/home/conda/feedstock_root/build_artifacts/ambertools_1714660402389/test_tmp/run_test.sh: line 50: sander.MPI: command not found
WARNING: Tests failed for ambertools-23.6-nompi_py312h5191d88_102.conda - moving package to /home/conda/feedstock_root/build_artifacts/broken
TESTS FAILED: ambertools-23.6-nompi_py312h5191d88_102.conda

##[error]Bash exited with code '1'.

**Comment by njzjz** [comment]
> ```
> 
> + sander --version
> 
> sander: Version 22.0
> 
> + sander.MPI --version
> 
> /home/conda/feedstock_root/build_artifacts/ambertools_1714660402389/test_tmp/run_test.sh: line 50: sander.MPI: command not found
> 
> WARNING: Tests failed for ambertools-23.6-nompi_py312h5191d88_102.conda - moving package to /home/conda/feedstock_root/build_artifacts/broken
> 
> TESTS FAILED: ambertools-23.6-nompi_py312h5191d88_102.conda
> 
> 
> 
> ##[error]Bash exited with code '1'.
> 
> 

These are nompi builds...

**Comment by mattwthompson** [comment]
Haha, that makes more sense

**Review by mattwthompson — APPROVED** [review]

**Comment by mattwthompson** [comment]
That's weird. Any ideas besides bringing down `openssh` ... ?

**Comment by njzjz** [comment]
I set `export OMPI_MCA_plm_ssh_agent=false` as suggested by https://github.com/conda-forge/openmpi-feedstock/issues/152#issuecomment-2066413862.

**Comment by njzjz** [comment]
> I set `export OMPI_MCA_plm_ssh_agent=false` as suggested by [conda-forge/openmpi-feedstock#152 (comment)](https://github.com/conda-forge/openmpi-feedstock/issues/152#issuecomment-2066413862).

Oh, it doesn't work :(

**Comment by mattwthompson** [comment]
Works well enough for us 🙃

Thanks for your work and patience @njzjz - I think people will like this addition

**Pull request merged**
